### PR TITLE
System.Diagnostics Tracing APIs (#35220)

### PR DESCRIPTION
**Pull request by tarekgh**
This change is porting the added System.Diagnostics APIs to preview 4. These APIs are very important to fill the gap between .NET Libraries and OpenTelemtry. There are some teams (application insight and Azure Monitoring) waiting to try the new APIs as early as we can provide it. It will be good if we can have these APIs included in preview 4 for early adaptors.

Original PR merged to master: https://github.com/dotnet/runtime/pull/35220
Original issue tracking the work:  #31373
The design issue for the APIs: dotnet/designs#98

CC @noahfalk

### CODE REVIEW

**Comment by tarekgh** [comment]
@ericstj @Anipik please let me know if there is anything else I should do to get this to preview 4. Thanks!

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+        }
+    }
+
+    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
+    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
+    // during the enumeration which force restarting the enumeration again.
+    // Causion: We can have the action executed on the same item more than once which is ok in our scenarios.
+    internal class SynchronizedList<T>
+    {
+        private List<T> _list;
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:256]:
```suggestion
        private readonly List<T> _list;
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+                    listeners.EnumWithAction(listener => listener.ActivityStopped?.Invoke(activity));
+                });
+            }
+        }
+    }
+
+    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
+    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
+    // during the enumeration which force restarting the enumeration again.
+    // Causion: We can have the action executed on the same item more than once which is ok in our scenarios.
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:253]:
```suggestion
    // Caution: We can have the action executed on the same item more than once which is ok in our scenarios.
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+                _list.Add(item);
+                _version++;
+            }
+        }
+
+        public bool AddIfNotExist(T item)
+        {
+            lock (_list)
+            {
+                if (!_list.Contains(item))
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:274]: How big do we expect these lists to be?  And more importantly, what's the worst case?  N calls to AddIfNotExist results in O(N^2) work here.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+            while (index < _list.Count)
+            {
+                T item;
+                lock (_list)
+                {
+                    if (version != _version)
+                    {
+                        version = _version;
+                        index = 0;
+                        continue;
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:313]: So if there's a mutation we start over? Hence the warning about potentially invoking delegates multiple times?  This means if one bad actor was repeatedly adding / removing / adding / removing / etc. while someone else was enumerating, that someone else might never end?

**Comment by tarekgh** [comment]
@stephentoub could you please tell which of your comments should block getting this to preview 4? I am asking because I can address your comments on master instead and safely catch preview 4.

**Comment by stephentoub** [comment]
I didn't realize this was already merged to master.

It'd be good to at least reason about the last two comments.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs (hunk 4 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:313]: Yes, it is safe in the calling scenarios. We are calling the listeners to detect if we need to create the Activity. I am not seeing a problem calling the listener multiple times with the same parameters multiple parameters when someone else mutates the list. 
The lists we have are the ActivtySource lists and listeners inside the activity source. If there is a bad actor you can do many more bad things creating a lot of Activity sources and listeners. or even in the listening can spend a long time in the execution.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs (hunk 3 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:274]: We are expecting the lists to be short. For activity sources, we expect would be one per component if the component publishing any tracing data. For the listeners, usually, the list will be shorter as in main scenario cases we expect one listener for who is interested to get the Activity notification (e.g. OpenTelemetry for instance). usual apps/Libraries not going to create listeners.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 5]
```diff
@@ -29,11 +30,14 @@ namespace System.Diagnostics
...
     /// </summary>
-    public partial class Activity
+    public partial class Activity : IDisposable
     {
 #pragma warning disable CA1825 // Array.Empty<T>() doesn't exist in all configurations
         private static readonly IEnumerable<KeyValuePair<string, string?>> s_emptyBaggageTags = new KeyValuePair<string, string?>[0];
+        private static readonly IEnumerable<ActivityLink> s_emptyLinks = new ActivityLink[0];
+        private static readonly IEnumerable<ActivityEvent> s_emptyEvents = new ActivityEvent[0];
 #pragma warning restore CA1825
+        private static ActivitySource s_defaultSource = new ActivitySource(string.Empty);
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:40]: I'll do that.

**Comment by danmosemsft** [comment]
Approved for P4, merging today would be ideal -- cc @Anipik

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 6]
```diff
@@ -70,8 +74,17 @@ public partial class Activity
...
-        private KeyValueListNode? _baggage;
+        private LinkedListNode<KeyValuePair<string, string?>>? _tags;
+        private LinkedListNode<KeyValuePair<string, string?>>? _baggage;
+        private LinkedListNode<ActivityLink>? _links;
+        private LinkedListNode<ActivityEvent>? _events;
+        private ConcurrentDictionary<string, object>? _customProperties;
+        private string? _displayName;
+
+        /// <summary>
+        /// Kind describes the relationship between the Activity, its parents, and its children in a Trace.
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:85]: I'll do that.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 7]
```diff
@@ -81,6 +94,23 @@ public partial class Activity
         /// </summary>
         public string OperationName { get; } = null!;
 
+        /// <summary>
+        /// DisplayName is name mainly intended to be used in the UI and not necessary has to be
+        /// same as OperationName.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:100]: I'll do that.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 8]
```diff
@@ -81,6 +94,23 @@ public partial class Activity
...
+        public string DisplayName
+        {
+            get => _displayName ?? OperationName;
+            set => _displayName = value;
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:104]: I'll change this to 
set => _displayName = value ?? throw new ArgumentNullException(nameof(value));

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 9]
```diff
@@ -81,6 +94,23 @@ public partial class Activity
...
+        }
+
+        /// <summary>
+        /// Get the ActivitySource object associated with this Activity.
+        /// All Activities created from public constructors will have a singleton source where the source name is empty string.
+        /// Otherwise, the source will be holding the object that created the Activity through ActivitySource.StartActivity.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:111]: I'll do that

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 10]
```diff
@@ -216,23 +246,69 @@ public partial class Activity
...
                     Iterate(tags) :
                     s_emptyBaggageTags;
 
-                static IEnumerable<KeyValuePair<string, string?>> Iterate(KeyValueListNode? tags)
+                static IEnumerable<KeyValuePair<string, string?>> Iterate(LinkedListNode<KeyValuePair<string, string?>>? tags)
                 {
                     do
                     {
-                        yield return tags!.keyValue;
+                        yield return tags!.Value;
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:258]: I'll do that in all places.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs [hunk 11]
```diff
@@ -718,6 +832,140 @@ private static bool IsW3CId(string id)
...
+            if (activity.IdFormat == ActivityIdFormat.W3C)
+                activity.GenerateW3CId();
+            else
+                activity._id = activity.GenerateHierarchicalId();
+
+            if (links != null)
+            {
+                foreach (ActivityLink link in links)
+                {
+                     activity._links = new LinkedListNode<ActivityLink>(link, activity._links);
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs:943]: tags existed before and the order was not important. But after getting OpenTelelmtery Specs, now it requires the order be preserved. I'll try to fix that to preserve the order. @noahfalk do you see any problem if we fixed that for the Tags?

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        public override bool Equals(object? obj) => (obj is ActivityContext context) ? Equals(context) : false;
+        public static bool operator ==(ActivityContext left, ActivityContext right) => left.Equals(right);
+        public static bool operator !=(ActivityContext left, ActivityContext right) => !(left == right);
+
+        public override int GetHashCode()
+        {
+            if (this == default)
+                return 0;
+
+            // HashCode.Combine would be the best but we need to compile for the full framework which require adding dependency
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs:67]: I am seeing it is not worth it for these cases. but I'll try to do it. will need to either introduce ifdef or split the files.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        /// </summary>
+        public ActivitySpanId SpanId { get; }
+
+        /// <summary>
+        /// The flags for the details about the trace.
+        /// </summary>
+        public ActivityTraceFlags TraceFlags { get; }
+
+        /// <summary>
+        /// system-specific configuration data.
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityContext.cs:52]: We can clean this up when docs pick them up.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityEvent.cs [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        /// Initializes a new instance of the <see cref="ActivityEvent"/> class.
+        /// </summary>
+        /// <param name="name">Event name.</param>
+        /// <param name="timestamp">Event timestamp. Timestamp MUST only be used for the events that happened in the past, not at the moment of this call.</param>
+        /// <param name="attributes">Event attributes.</param>
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+        {
+            this.Name = name ?? string.Empty;
+            this.Attributes = attributes ?? s_emptyAttributes;
+            this.Timestamp = timestamp != default ? timestamp : DateTimeOffset.UtcNow;
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityEvent.cs:56]: Yes, you can see we already doing that in Activity class 

```C#
            if (StartTimeUtc == default)
                StartTimeUtc = GetUtcNow();
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityEvent.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ActivityEvent"/> class.
+        /// </summary>
+        /// <param name="name">Event name.</param>
+        /// <param name="attributes">Event attributes.</param>
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        {
+            this.Name = name ?? string.Empty;
+            this.Attributes = attributes ?? s_emptyAttributes;
+            this.Timestamp = DateTimeOffset.UtcNow;
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityEvent.cs:43]: No specific reason. I'll do that.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityEvent.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+        /// <summary>
+        /// Gets the <see cref="ActivityEvent"/> timestamp.
+        /// </summary>
+        public DateTimeOffset Timestamp { get; }
+
+        /// <summary>
+        /// Gets the collection of attributes associated with the event.
+        /// </summary>
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityEvent.cs:72]: I think you are right. Attributes shouldn't be nullable here. Thanks.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityLink.cs [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        public bool Equals(ActivityLink value) => Context == value.Context && value.Attributes == Attributes;
+        public static bool operator ==(ActivityLink left, ActivityLink right) => left.Equals(right);
+        public static bool operator !=(ActivityLink left, ActivityLink right) => !left.Equals(right);
+
+        public override int GetHashCode()
+        {
+            if (this == default)
+                return 0;
+
+            // HashCode.Combine would be the best but we need to compile for the full framework which require adding dependency
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityLink.cs:55]: Will try to do that.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityListener.cs [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    /// ActivityListener allows listening to the start and stop Activity events and give the oppertunity to decide creating the Activity for sampling scenarios.
+    /// </summary>
+    public sealed class ActivityListener : IDisposable
+    {
+        /// <summary>
+        /// Construct a new <see cref="ActivityListener"/> object to start listeneing to the <see cref="Activity"/> events.
+        /// </summary>
+        public ActivityListener()
+        {
+
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityListener.cs:24]: will remove it,

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+
+using System.Threading;
+using System.Collections.Generic;
+
+namespace System.Diagnostics
+{
+    public sealed class ActivitySource : IDisposable
+    {
+        private static SynchronizedList<ActivitySource> s_activeSources = new SynchronizedList<ActivitySource>();
+        private static SynchronizedList<ActivityListener> s_allListeners = new SynchronizedList<ActivityListener>();
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:13]: will do that.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+        private SynchronizedList<ActivityListener>? _listeners;
+
+        private ActivitySource() { throw new InvalidOperationException(); }
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:16]: I left it when I was doing some internal testing. It was not intended to stay. Thanks for the catch.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:256]: will do it.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:253]: why English lack having `Causion` :-) interesting the editor spell checker didn't complain. of course, I'll fix it. thanks.

**Comment by tarekgh** [comment]
Talked to @stephentoub offline and learned none of the feedback comments is P4 blocker. So, I'll go ahead and merge this one and will address the feedback on the master. Thanks, @stephentoub for the feedback.

**Pull request merged**